File: packages/marts/deno_dist/habitica.js
// habitica.ts
var opts = {
  "habitica.api_url": {
    env_var_name: "PPC_HABITICA_API_URL",
    default_val: "https://habitica.com/api/v3"
  },
  "habitica.user_id": {
    env_var_name: "PPC_HABITICA_USER_ID"
  },
  "habitica.client_name": {
    default_val: "3544a0b8-d71a-46e0-9bb1-6ddbb2abcddb-PPC-Software"
  },
  "habitica.api_token": {
    env_var_name: "PPC_HABITICA_API_TOKEN"
  }
};
async function habitica(mize) {
  console.log("hiiiiiiiiiiiiiiiiiiii");
  mize.add_opts(opts);
  mize.add_part(new Habitica(mize));
}
async function handleRateLimit(response) {
  const limit = response.headers.get("X-RateLimit-Limit") || "NONE";
  const remaining = parseInt(response.headers.get("X-RateLimit-Remaining") || "10", 10);
  const reset = response.headers.get("X-RateLimit-Reset");
  console.log(`RateLimit: ${limit} | Remaining: ${remaining} | Reset: ${reset}`);
  if (remaining < 2 && reset) {
    const resetDate = new Date(reset);
    const now = /* @__PURE__ */ new Date();
    const waitMs = resetDate.getTime() - now.getTime() + 1e3;
    if (waitMs > 0) {
      console.log(`Waiting ${Math.round(waitMs / 1e3)} secs for next rate limit window...`);
      await new Promise((resolve) => setTimeout(resolve, waitMs));
    }
  }
}
var Habitica = class {
  mize;
  constructor(mize) {
    this.mize = mize;
  }
  async get_tasks(type = "todos") {
    return await this.api_request("GET", `tasks/user?type=${type}`);
  }
  async delete_task(id) {
    await this.api_request("DELETE", `tasks/${id}`);
  }
  async api_request(method, path, extraHeaders = {}, data = {}) {
    const mize = this.mize;
    if (!path.startsWith("/")) {
      path = "/" + path;
    }
    const headers = Object.assign({
      "Content-Type": "application/json",
      "x-api-user": mize.get_config("habitica.user_id"),
      "x-api-key": mize.get_config("habitica.api_token"),
      "x-client": mize.get_config("habitica.client_name")
    }, extraHeaders);
    const response = await fetch(mize.get_config("habitica.api_url") + path, {
      method,
      headers,
      body: JSON.stringify(data)
    });
    await handleRateLimit(response);
    if (!response.ok) {
      console.log("request_url:", mize.get_config("habitica.api_url") + path);
      throw new Error(`Habitica API error: ${response.status} ${response.statusText}`);
    }
    return (await response.json()).data;
  }
};
export {
  habitica,
  opts
};

File: packages/marts/habitica.rs
use chrono::{DateTime, Utc};
use mize::{mize_err, mize_part, Mize, MizeError, MizePart, MizeResult};
use reqwest::{blocking::Client, blocking::Response, header, Method};
use serde_json::{json, Value};
use std::thread::sleep;
use std::time::Duration;

#[mize_part]
#[derive(Default)]
pub struct Habitica {
    mize: Mize,
    client: Client,
}

pub fn habitica(mize: &mut Mize) -> MizeResult<()> {
    let client = Client::new();
    mize.register_part(Box::new(Habitica {
        mize: mize.clone(),
        client,
    }))
}

impl MizePart for Habitica {
    fn opts(&self, mize: &mut Mize) {
        mize.new_opt("habitica.api_url");
        mize.new_opt("habitica.user_id");
        mize.new_opt("habitica.api_token");
        mize.new_opt("habitica.client_name");
    }
}

impl Habitica {
    pub fn api_request(&mut self, method: Method, path: String, data: Value) -> MizeResult<Value> {
        let api_url = self.mize.get_config("habitica.api_url")?.to_string();
        let user_id = self.mize.get_config("habitica.user_id")?.to_string();
        let api_token = self.mize.get_config("habitica.api_token")?.to_string();
        let client_name = self.mize.get_config("habitica.client_name")?.to_string();

        let mut headers = header::HeaderMap::new();
        headers.insert(
            "Content-Type",
            header::HeaderValue::from_static("application/json"),
        );
        headers.insert("x-api-user", header::HeaderValue::from_str(&user_id)?);
        headers.insert("x-api-key", header::HeaderValue::from_str(&api_token)?);
        headers.insert("x-client", header::HeaderValue::from_str(&client_name)?);

        let url = format!("{}/{}", api_url, path.trim_start_matches('/'));

        let mut request_builder = self.client.request(method, &url).headers(headers);

        if data != json!({}) {
            request_builder = request_builder.json(&data);
        }

        let response = request_builder.send()?;

        handle_rate_limit(&response);

        if !response.status().is_success() {
            let status = response.status();
            let text = response.text().unwrap_or_default();
            println!("request_url: {}", url);
            return Err(mize_err!(
                "Habitica API error: {} {} - {}",
                status.as_u16(),
                status.canonical_reason().unwrap_or(""),
                text
            ));
        }

        let json_response: Value = response.json()?;
        Ok(json_response.get("data").cloned().unwrap_or(Value::Null))
    }

    pub fn get_tasks(&mut self, task_type: &str) -> MizeResult<Value> {
        self.api_request(
            Method::GET,
            format!("tasks/user?type={}", task_type),
            json!({}),
        )
    }

    pub fn delete_task(&mut self, id: &str) -> MizeResult<Value> {
        self.api_request(Method::DELETE, format!("tasks/{}", id), json!({}))
    }
}

fn handle_rate_limit(response: &Response) {
    let headers = response.headers();
    let limit = headers
        .get("X-RateLimit-Limit")
        .and_then(|v| v.to_str().ok())
        .unwrap_or("NONE");
    let remaining = headers
        .get("X-RateLimit-Remaining")
        .and_then(|v| v.to_str().ok())
        .and_then(|s| s.parse::<i32>().ok())
        .unwrap_or(10);
    let reset = headers
        .get("X-RateLimit-Reset")
        .and_then(|v| v.to_str().ok());

    println!(
        "RateLimit: {} | Remaining: {} | Reset: {}",
        limit,
        remaining,
        reset.unwrap_or("N/A")
    );

    if remaining < 2 {
        if let Some(reset_str) = reset {
            if let Ok(reset_date) = DateTime::parse_from_rfc2822(reset_str) {
                let now = Utc::now();
                let wait_duration = reset_date.signed_duration_since(now).to_std();
                if let Ok(wait_duration) = wait_duration {
                    let wait_ms = wait_duration.as_millis() as u64 + 1000;
                    println!(
                        "Waiting {} secs for next rate limit window...",
                        (wait_ms / 1000)
                    );
                    sleep(Duration::from_millis(wait_ms));
                }
            }
        }
    }
}
